### ISSUE TITLE
Intellisense is gone (again)

### ISSUE THREAD

**Issue by Skyedra** [dotnet/vscode-csharp#6024]
Type: <b>Bug</b>

Definitely related to the latest crop of issues:  https://github.com/microsoft/vscode-dotnettools/issues/317

Not sure what happened; saw some c# update fly past when I booted up today and now intellisense doesn't work.  I don't see anything terribly relevant like I normally would in the logs when autoupdate breaks something (like roslyn load error, omnisharp load error, etc).  

This is like the fifth time an autoupdate has broken intellisense for me, so this is getting to be pretty frustrating.  I don't need my IDE to autoupdate, I just need it to work.  Gonna try to figure out how to stop autoupdates now.  I respect it's a free tool, but when I'm down like this there's a real financial impact to me that doesn't make it worth me being microsoft's unpaid QA.  Sorry for frustration.

Extension version: 2.0.320
VS Code version: Code 1.79.2 (695af097c7bd098fbf017ce3ac85e09bbc5dda06, 2023-06-14T08:59:55.818Z)
OS version: Linux x64 6.2.0-26-generic
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|13th Gen Intel(R) Core(TM) i7-13700K (24 x 5300)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>video_decode: enabled<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|1, 1, 1|
|Memory (System)|62.55GB (48.87GB free)|
|Process Argv|--unity-launch|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|plasma|
|XDG_CURRENT_DESKTOP|KDE|
|XDG_SESSION_DESKTOP|KDE|
|XDG_SESSION_TYPE|x11|
</details>

**Comment by Skyedra** [comment]
FYI reverting to v1.26.0 seems to fix

**Comment by Skyedra** [comment]
Log comparison, may be useful.

Old version (intellisense working):   C# tab empty, but this is shown in omnisharp log:

```

Starting OmniSharp server at 8/3/2023, 5:28:07 PM
    Target: /project-path-redacted/core

OmniSharp server started with .NET 6.0.120
.
    Path: /home/skye/.vscode/extensions/ms-dotnettools.csharp-1.26.0-linux-x64/.omnisharp/1.39.7-net6.0/OmniSharp.dll
    PID: 185343

[info]: OmniSharp.Stdio.Host
        Starting OmniSharp on Unknown 0.0 (Unknown)
[info]: OmniSharp.Services.DotNetCliService
        Checking the 'DOTNET_ROOT' environment variable to find a .NET SDK
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to /usr/lib/dotnet/dotnet
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 1 MSBuild instance(s)
            1: .NET Core SDK 6.0.120 17.0.1 - "/usr/lib/dotnet/sdk/6.0.120/"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: .NET Core SDK 6.0.120 17.0.1 - "/usr/lib/dotnet/sdk/6.0.120/"
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
[info]: OmniSharp.Cake.CakeProjectSystem
        Detecting Cake files in '/project-path-redacted/core'.
[info]: OmniSharp.Cake.CakeProjectSystem
        Did not find any Cake files
[info]: OmniSharp.MSBuild.ProjectSystem
        No solution files found in '/project-path-redacted/core'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for '/project-path-redacted/core/core.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for '/project-path-redacted/core/src/Org/AStar/Roy-T.AStar/Roy-T.AStar.csproj'
[info]: OmniSharp.Script.ScriptProjectSystem
        Detecting CSX files in '/project-path-redacted/core'.
[info]: OmniSharp.Script.ScriptProjectSystem
        Did not find any CSX files
[info]: OmniSharp.WorkspaceInitializer
        Configuration finished.
[info]: OmniSharp.Stdio.Host
        Omnisharp server running using Stdio at location '/project-path-redacted/core' on host 184753.
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: /project-path-redacted/core/core.csproj
[info]: OmniSharp.MSBuild.ProjectManager
        Successfully loaded project file '/project-path-redacted/core/core.csproj'.
[info]: OmniSharp.MSBuild.ProjectManager
        Adding project '/project-path-redacted/core/core.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: /project-path-redacted/core/src/Org/AStar/Roy-T.AStar/Roy-T.AStar.csproj
[info]: OmniSharp.MSBuild.ProjectManager
        Successfully loaded project file '/project-path-redacted/core/src/Org/AStar/Roy-T.AStar/Roy-T.AStar.csproj'.
[info]: OmniSharp.MSBuild.ProjectManager
        Adding project '/project-path-redacted/core/src/Org/AStar/Roy-T.AStar/Roy-T.AStar.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Update project: core
[info]: OmniSharp.MSBuild.ProjectManager
        Found referenced project outside root directory: /project-path-redacted/CommonSharp/CommonSharp.csproj
[info]: OmniSharp.MSBuild.ProjectManager
        Adding project '/project-path-redacted/CommonSharp/CommonSharp.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for '/project-path-redacted/CommonSharp/CommonSharp.csproj'
[info]: OmniSharp.MSBuild.ProjectManager
        Update project: Roy-T.AStar
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: /project-path-redacted/CommonSharp/CommonSharp.csproj
[info]: OmniSharp.MSBuild.ProjectManager
        Successfully loaded project file '/project-path-redacted/CommonSharp/CommonSharp.csproj'.
[info]: OmniSharp.MSBuild.ProjectManager
        Update project: CommonSharp
[info]: OmniSharp.OmniSharpWorkspace
        Adding analyzer reference: /usr/lib/dotnet/sdk/6.0.120/Sdks/Microsoft.NET.Sdk/analyzers/Microsoft.CodeAnalysis.CSharp.NetAnalyzers.dll
[info]: OmniSharp.OmniSharpWorkspace
        Adding analyzer reference: /usr/lib/dotnet/sdk/6.0.120/Sdks/Microsoft.NET.Sdk/analyzers/Microsoft.CodeAnalysis.NetAnalyzers.dll
[info]: OmniSharp.OmniSharpWorkspace
        Adding analyzer reference: /usr/lib/dotnet/packs/Microsoft.NETCore.App.Ref/6.0.20/analyzers/dotnet/cs/System.Text.Json.SourceGeneration.dll
[info]: OmniSharp.Roslyn.CSharp.Services.Diagnostics.CSharpDiagnosticWorkerWithAnalyzers
        Solution initialized -> queue all documents for code analysis. Initial document count: 520.
Received response for /v2/getcodeactions but could not find request.
[warn]: OmniSharp.Roslyn.CSharp.Services.Navigation.FindUsagesService
        No symbol found. File: /project-path-redacted/core/src/Underworld/Modding/ModSpeciesFacade.cs, Line: 179, Column: 13.

```

New version (intellisense broken):  No omnisharp anymore, but this is C# tab:

```
Dotnet path: /home/skye/.config/Code/User/globalStorage/ms-dotnettools.vscode-dotnet-runtime/.dotnet/7.0.9/dotnet
Activating C# standalone...
info: LanguageServerHost[0]
      Starting server...
[Info  - 5:36:04 PM] [LanguageServerProjectSystem]Loaded MSBuild at /usr/lib/dotnet/sdk/6.0.120
[Info  - 5:36:04 PM] [LanguageServerProjectSystem]Loading /project-path-redacted/core/src/Org/AStar/Roy-T.AStar.sln...
[Warn  - 5:36:04 PM] [LanguageServerProjectSystem]Failure while loading /project-path-redacted/core/src/Org/AStar/Roy-T.AStar.Benchmark/Roy-T.AStar.Benchmark.csproj: Could not find a part of the path '/project-path-redacted/core/src/Org/AStar/Roy-T.AStar.Benchmark/Roy-T.AStar.Benchmark.csproj'.
[Warn  - 5:36:04 PM] [LanguageServerProjectSystem]Failure while loading /project-path-redacted/core/src/Org/AStar/Roy-T.AStar.Tests/Roy-T.AStar.Tests.csproj: Could not find a part of the path '/project-path-redacted/core/src/Org/AStar/Roy-T.AStar.Tests/Roy-T.AStar.Tests.csproj'.
[Warn  - 5:36:04 PM] [LanguageServerProjectSystem]Failure while loading /project-path-redacted/core/src/Org/AStar/Roy-T.AStar.Viewer/Roy-T.AStar.Viewer.csproj: Could not find a part of the path '/project-path-redacted/core/src/Org/AStar/Roy-T.AStar.Viewer/Roy-T.AStar.Viewer.csproj'.
[Info  - 5:36:04 PM] [LanguageServerProjectSystem]Successfully completed load of /project-path-redacted/core/src/Org/AStar/Roy-T.AStar/Roy-T.AStar.csproj
[Info  - 5:36:04 PM] [LanguageServerProjectSystem]Completed (re)load of all projects in 00:00:00.4611082
```

New version doesn't seem to try to load all loaded projects, just the first one in the workspace?

**Comment by eduardoscheffer** [comment]
![csharp-extension-bug](https://github.com/dotnet/vscode-csharp/assets/107814795/cc2a327c-4825-44df-82be-8bdfff3c382c)

FYI reverting to v1.26.0

**Comment by ZhaoYuNing123** [comment]
> ![csharp-extension-bug](https://user-images.githubusercontent.com/107814795/258279863-cc2a327c-4825-44df-82be-8bdfff3c382c.jpg)
> 
> FYI reverting to v1.26.0

how to revert?

**Comment by hongnv55** [comment]
@ZhaoYuNing123 It's easy by install another version
![ảnh](https://github.com/dotnet/vscode-csharp/assets/1445342/b5a374d6-4e61-4422-92ed-75918a126766)

**Comment by BrianCArnold** [comment]
Can I ask why "break thousands of people's workflows with absolutely no warning or explanation" was on the roadmap for Visual Studio Code this month?

**Comment by bbuchanan** [comment]
Having the same problem, reverting to 1.26 is a workaround.

**Comment by jasonmalinowski** [comment]
@Skyedra: thanks for the logs, that's very helpful!

The C# extension will be loading all of the projects listed in the solution file. Looking at your two sets of logs, I'm guessing it means you have projects that _aren't_ in the solution file (like your core.csproj), but it also looks like your solution file has some things that aren't being loaded by Omnisharp (like it's trying to load Roy-T.AStar.Viewer.csproj, but can't find it). Is your solution file out of date with what is otherwise in your source folders? Do the projects it's failing to find actually exist?

**Comment by JustinGrote** [comment]
Revert to 1.26 worked for me too. I don't have a solution file, only a .csproj, is that going to be required going forward, if so, why? It worked just fine before.

**Comment by Skyedra** [comment]
@jasonmalinowski  I'm not using a .sln for any of my code on the project (I'm not actually familiar with generating one outside visual studio classic, and thought that was just a legacy import thing), but I do have a .code-workspace:

```
{
	"folders": [
		{
			"path": "core"
		},
		{
			"path": "hatch"
		},
		{
			"path": "Switchboard"
		},
		{
			"path": "Bridge"
		},
		{
			"path": "Supervisor"
		},
		{
			"path": "LoadTester"
		},
		{
			"path": "CommonSharp"
		}
	],
	"settings": {
		"omnisharp.useModernNet": true,
		"omnisharp.enableRoslynAnalyzers": true
	}
}
```

Pretty sure the two settings were workarounds from previous omnisharp autoupdate breakage, though as omnisharp isn't used anymore I imagine those settings wouldn't affect anything.

Looking at the core.csproj in particular (where astar lives inside of (src/Org/Astar, next to src/ProjectName), there is a `Roy-T.AStar.sln` in the directory tree but it's not used.  (It's a library that got pulled into the source tree path directly, I'm guessing because project-specific edits had to be made at some point)

core.csproj:

```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
	 <OutputType>Library</OutputType>
	 <DefineConstants Condition="'$(SET_PRIVBUILD)'=='true'">$(DefineConstants);PRIVBUILD</DefineConstants>
	 <DefineConstants Condition="'$(SET_LOCALGAMETEXTPRIVATE)'=='true'">$(DefineConstants);LOCALGAMETEXTPRIVATE</DefineConstants>
	 <!-- Following two lines appear to fix a "Error CS0579 Duplicate 'global::System.Runtime.Versioning.TargetFrameworkAttribute'" that suddenly showed up one day -->
	 <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
	 <GenerateTargetFrameworkAttribute>false</GenerateTargetFrameworkAttribute> 
  </PropertyGroup>

  <ItemGroup>
	<Content Include="GameText.xml" CopyToPublishDirectory="PreserveNewest" />
	<Content Condition="'$(SET_LOCALGAMETEXTPRIVATE)'=='true'" Include="GameTextPrivate.xml" CopyToPublishDirectory="PreserveNewest" />  <!-- TODO: Don't include on public build, if it matters.-->
	<None Include="maps/**" CopyToPublishDirectory="PreserveNewest" LinkBase="maps/" />
	<None Include="shops/**" CopyToPublishDirectory="PreserveNewest" LinkBase="maps/" />
	<None Include="testSaves/**" CopyToPublishDirectory="PreserveNewest" LinkBase="testSaves/" />
	<None Include="procedural/**" CopyToPublishDirectory="PreserveNewest" LinkBase="procedural/" />
	<Content Include="OldSaveImportRenameMap.txt" CopyToPublishDirectory="PreserveNewest" />
  </ItemGroup>

  <ItemGroup>
    <PackageReference Include="MoonSharp" Version="2.0.0" />
    <PackageReference Include="WeightedRandomizer" Version="2.0.1" />
    <PackageReference Include="ZstdNet" Version="1.4.5" />
    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />	
  </ItemGroup>

  <ItemGroup>
	<ProjectReference Include="..\CommonSharp\CommonSharp.csproj" />
  </ItemGroup>
</Project>
```

So it is a bit of a mystery why it's pulling an unused/unreferenced .sln file in a deep subfolder given the .csproj for that subproject doesn't reference it.  And then ignores the src folder entirely.

**Comment by Skyedra** [comment]
Also, project tree if it helps:
```
find . | grep -P "(.*)(csproj|sln|code-workspace)$"
./hatch/hatch.csproj
./as3tocs/as3tocs.csproj
./Bridge/Bridge.csproj
./CommonSharp/CommonSharp.csproj
./Supervisor/Supervisor.csproj
./LoadTester/LoadTester.csproj
./core/src/Org/AStar/Roy-T.AStar.sln
./core/src/Org/AStar/Roy-T.AStar/Roy-T.AStar.csproj
./core/core.csproj
./Switchboard/Switchboard.csproj
./ProjectNameHere.code-workspace
```

(as3tocs is not actively used, so it's not in the code workspace)

**Comment by jasonmalinowski** [comment]
@Skyedra: That is very helpful to have the project tree, so thanks. Just answering your comments in order:

> Pretty sure the two settings were workarounds from previous omnisharp autoupdate breakage, though as omnisharp isn't used anymore I imagine those settings wouldn't affect anything.

Yeah you won't need those if Omisharp isn't active. (and generally you shouldn't need them going forward regardless...)

> So it is a bit of a mystery why it's pulling an unused/unreferenced .sln file in a deep subfolder given the .csproj for that subproject doesn't reference it. And then ignores the src folder entirely.

The v2 extension more aggressively prefers the solution file, or at least our assumption was if there _was_ a solution file, you wanted that used. Did you have `omnisharp.enableMsBuildLoadProjectsOnDemand` set in your user options?

**Comment by Skyedra** [comment]
Thanks for the info; a couple thoughts so breaking this into sections:

### User prefs:

I don't believe I have that setting in user prefs.  Here's what I do have in settings.json:

```
{
    "telemetry.enableCrashReporter": false,
    "workbench.sideBar.location": "left",
    "editor.insertSpaces": false,
	"editor.detectIndentation": false,
	"debug.openExplorerOnEnd": true,
	"debug.console.closeOnEnd": true,
	//"omnisharp.path": "latest",
	//"omnisharp.path": "1.37.17",
	"debug.onTaskErrors": "showErrors",
	"editor.rulers": [
		120
	],
	"debug.openDebug": "openOnSessionStart",
	"diffEditor.ignoreTrimWhitespace": false
}
```

### Prioritization of solution files vs csproj files:

I don't think prioritizing a solution file deep within the subtree is ideal.  Also odd to ignore a csproj in the root of the subproject folder over a sln in a deeper folder.  But more generally, I would definitely prefer trying to stick as close to the compiler logic as possible.  

My expectation would be something like:

1. For each project in workspace
   A. Use whatever logic `dotnet build` would do in that folder to determine project scope

(I respect due to being different tools, this simplification is perhaps _much_ easier said than done.  Just selecting based on there being a csproj or sln in the subproject's root folder would probably solve most cases, including my own)

Also, if only one subproject at a time can be active (as in omnisharp), I can live with the intellisense project selector at the bottom like omnisharp had.  However, being able to not deal with that fiddling would be a major feature worth upgrading for.

### User confusion on solution files in vscode:  

My *impression* as a user (which may not be accurate!) is that sln's are legacy and have been replaced by the set of smaller/user-readable files like csproj/code-workspace/launch.json/etc.  My impression is they were mostly intended for easily migrating old projects from Visual Studio -> VS Code and not for using in new projects.

I'm not deeply opposed to solution files, but I haven't used them since c++ days so they carry a certain amount of, uh, emotional baggage?  I wouldn't want to have to manually add each cs file to a solution file, and manually keep directory tree vs solution tree filenames synchronized, etc.  Also, I seemingly can't manage sln with the IDE, and I recall them being a PITA to do [by hand](https://learn.microsoft.com/en-us/visualstudio/extensibility/internals/solution-dot-sln-file?view=vs-2022)... so all that is to say, I'm not much inclined to switch to using sln files at first glance.  I do see there is some functionality with [dotnet sln](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-sln) now, and I may not need to manage it constantly like in the old days, so I could potentially go that route.  

I also don't understand well the relationship between solution files and the new set of configs.  IE, would I use a solution file instead of a workspace file?  Instead of a workspace + csproj + launch files?

So, that is all a way of saying that if the legacy behavior cannot be retained then the upgrade should be opt-in and provide clear migration guide.

**Comment by navaddins** [comment]
Not sure it's bug or design. I'm same page with @JustinGrote. I'm using it for Azure function and there has no solutions. Suddenly, all Intellisense is gone. I didn't realize that's because of latest update. Uninstall/Install VS Code and uninstall/install C# for Visual Studio Code again but not luck. 😣Finally, I decided to use v1.26.0. It's working back as normal.

**Comment by Kidsunbo** [comment]
Install C# dev kit will fix all the problem. But I don't want this. Is this update a strategy to force us to use C# dev kit?

**Comment by egbertn** [comment]
> FYI reverting to v1.26.0 seems to fix

That's really old. We have 1.81 today.

**Comment by egbertn** [comment]
My suspect is that VS Code installs an old version of the .net runtime. 7.0.0-preview.7.22375.6.

While I have 7.0.9. 
This causes a defunct of  intellisense.

**Comment by johannesergon** [comment]
on our company, we are using a workspace with a lot of repositores (like a mono repo) which every repository contains one or multiple sln/csproj files and also a code-workspace file

for example my directory structure looks like this:
```
~/workspace/overall.code-workspace
~/workspace/repo1/project1/(project1.sln|project1.code-workspace)
~/workspace/repo1/project2/(project2.sln|project2.code-workspace)
~/workspace/repo2/project3/(project3.sln|project3.code-workspace)
~/workspace/repo3/project4/(project4.sln|project4.code-workspace)
~/workspace/repo3/project4/subproject1/subproject1.csproj
~/workspace/repo3/project4/subproject2/subproject2.csproj
```
and i'm opening `~/workspace/overall.code-workspace` with vscode which includes all global configuration
and then i'm switching between sln files depending on which project/repository i'm working on 

in this way its easier for me to search for source code accross all projects and repositories

and after updating to the version from friday the console output also stopped at
`standalone... info: LanguageServerHost[0] Starting server... `

**Comment by egbertn** [comment]
SOLUTION:
No downgrading, just a setting which disables dependency on the crappy language server.

How to use OmniSharp?
If you don’t want to take advantage of the great Language Server features, you can revert back to using OmniSharp by going to the Extension settings and setting dotnet.server.useOmnisharp to true. Next, uninstall or disable C# Dev Kit. Finally, restart VS Code for this to take effect.

**Comment by AdamRussak** [comment]
> ninstall or disable C# Dev Kit.

this is same as downgrade basically.

i hope they will fix it soon :(

**Comment by jasonmalinowski** [comment]
Going to comment to a few of you folks at once here, so my apologies if this gets a bit confusing:

> @johannesergon: and then i'm switching between sln files depending on which project/repository i'm working on

I can't quite tell, based on your comments, is the current experience working well for you or are you running into friction here?

> @Skyedra: Also, if only one subproject at a time can be active (as in omnisharp), I can live with the intellisense project selector at the bottom like omnisharp had

Just so I can understand your workflow a bit better, once you have VS Code open are you then using the switcher to switch between the project you're working on? Because looking at the Omnisharp behavior my gut is it'd still prefer the solution file (if that was mapped into your workspace), but then of course you could use the picker to change.

> @Kidsunbo: Install C# dev kit will fix all the problem. But I don't want this. Is this update a strategy to force us to use C# dev kit?

Absolutely not! One challenge with the existing Omnisharp project loading logic is it supports a lot of different scenarios and a lot of different workflows, so it proves to be a challenge to make sure we get all the cases right.

**Comment by JustinGrote** [comment]
@jasonmalinowski by my testing, these workflows were broken by the 2.0 (release) - And please confirm 2.0 release was intentional per https://github.com/dotnet/vscode-csharp/issues/6040 because it seems to me like an accidental prerelease was made public since there's no Github release:

- If you don't have a .sln (which is very common), intellisense seems to be broken
- If you don't have dev kit installed (because that's a licensed component), intellisense seems to be broken

Both of these assume you haven't checked the "use legacy omnisharp", though if you have and you have C# dev kit installed, it is ignored (as stated in the documentation)

**Comment by Skyedra** [comment]
@jasonmalinowski: 
> Just so I can understand your workflow a bit better, once you have VS Code open are you then using the switcher to switch between the project you're working on? Because looking at the Omnisharp behavior my gut is it'd still prefer the solution file (if that was mapped into your workspace), but then of course you could use the picker to change.

For better or worse, that is indeed my workflow.  😅  It is a bit of a hassle to have to switch using the picker every time I change subproject, but it works (or at least, used to).  I'm not aware of any way to get around this in vs code so I just put up with it.

![image](https://github.com/dotnet/vscode-csharp/assets/22365940/4030e1e9-8ce5-4c28-bbad-01f693837d04)

It's no big deal if it defaults to the solution file on first load but ideally it at least remembers the choice between sessions.

**Comment by jasonmalinowski** [comment]
Alright general status update for everyone: https://github.com/dotnet/vscode-csharp/issues/5923 is being used to track the failure to load csproj files if there's not a .sln file; a fix is on its way for that to at least load all .csproj files in the workspace if there is no .sln file on open. That doesn't give _quite_ the control that Omnisharp did but this should at least unblock many folks.

@Skyedra: in your case do you need the .sln file at all? Once the fix I mentioned for #5923 is out, if you were to delete the solution file I'd expect it to just load all your .csprojs and you won't have to switch between projects anymore. I admit I'm not entirely sure why you're using a workspace w/ folders in the way you are -- that might have been causing troubles for Omnisharp as well with that picker too, since otherwise it would have just offered an option to load all projects at once.

**Comment by jasonmalinowski** [comment]
And as a general request: if you have an issue with the v2 extension that is unrelated #5923 then I think at this point I'd request you file a new issue than comment here, only so that way we can make sure we capture all the different workflows well.

**Comment by Skyedra** [comment]
> in your case do you need the .sln file at all? 

I don't need the .sln and could delete it.  

That said, it doesn't seem completely unusual for people bringing in external code into subdirectories of their project tree to still have vestigial sln files they forget about.  The sln prioritization behavior is not intuitive, so I don't think this solution will be obvious to people.

**Comment by jasonmalinowski** [comment]
@Skyedra Agreed, it's not helpful in your case. So I'm getting your workflow right, a workspace setting to ignore the .sln and just load _all_ .csproj already mapped in your workspace (not just a folder in the workspace) would be best? That way all your projects load and no fiddling any more with the picker?

**Comment by Skyedra** [comment]
> a workspace setting to ignore the .sln and just load all .csproj already mapped in your workspace (not just a folder in the workspace) 

Yes, this would be quite handy.  

In terms of best solution, I did have one idea for something that *might* work even better?  Having toggleable checks on the picker would probably be the most flexible for peoples various usecases.  Like this kinda thing:

![image](https://github.com/dotnet/vscode-csharp/assets/22365940/c41ca679-18cd-4c86-af27-9eb45c96aaf5)

This would also be more clear to a new user why it's not loading intellisense instead of having to know about a specific workspace option.

**Comment by johannesergon** [comment]
> > @johannesergon: and then i'm switching between sln files depending on which project/repository i'm working on
> 
> I can't quite tell, based on your comments, is the current experience working well for you or are you running into friction here?
@jasonmalinowski 

```
~/workspace$ find . -type f -name "*.csproj" | wc -l
133
```
😅

For me its a good think to select only the projects which should be loaded, otherwise my my laptop will get super slow. With the omnisharp project/sln selector i can achive this.
I only have to select the correct sln file and wait for omnisharp to load everythink and than i can work normally 🙂

**Comment by paulinventome** [comment]
I can add my voice to this. Some updates happened and intellisense stopped working (actually it all lit up in nice colours and then they were all removed and not working). I did the downgrade and everything but didn't work. Also none of the SDK intellisense was working either, not just my own code. I migrated everything over to net7 (from 6) thinking it was related to this (the new c# seemed to want 7) and it did make a small difference. I find this thread and I think it's the same

I have a workspace with 4 folders added. Each of the folders is a separate project. I've found that intellisense works for the first project (the one that is launched) but the other 3 don't work. So clearly this is the issue described above, the language server isn't finding these projects.

I didn't think vscode worked with sln files, hence workspaces.

So what *is* the correct way to set up a vs code project with various class libraries in so that the language server can find them all?

**Comment by JustinGrote** [comment]
@paulinventome at the moment, do you have a solution file? Until a fix is in place, basically you do this:

```
dotnet new sln
dotnet sln add <path to .csproj folder>
```

Repeat above and then reload your vscode window and then it should be able to find them.

I have no idea about multi-root workspaces and any demons that might exist there, haven't tried it lately.

**Comment by sharwell** [comment]
> That said, it doesn't seem completely unusual for people bringing in external code into subdirectories of their project tree to still have vestigial sln files they forget about. The sln prioritization behavior is not intuitive, so I don't think this solution will be obvious to people.

I 100% agree here. In general, I would only expect it to locate a solution file by default if it's in the actual folder that was opened in VS Code. Unfortunately, I can think of multiple concrete cases where this restriction would start to break yet other workflows that people have.

The key is certainly going to be identifying and documenting the cases that people are expecting to work. It's often fairly straightforward to implement the behavior once it's clearly articulated. 😄

**Comment by paulinventome** [comment]
> @paulinventome at the moment, do you have a solution file? Until a fix is in place, basically you do this:
> 
> ```
> dotnet new sln
> dotnet sln add <path to .csproj folder>
> ```
> 
> Repeat above and then reload your vscode window and then it should be able to find them.
> 
> I have no idea about multi-root workspaces and any demons that might exist there, haven't tried it lately.

For the avoidance of doubt are we placing a solution file beside each project file or one solution file at the workspace folder with 4 x dotnet sln add <path to .csproj folder> applied?

thanks

**Comment by pyrrho** [comment]
> It's often fairly straightforward to implement the behavior once it's clearly articulated.

Articulating the odd cases that current users rely on can be much easier said than done, though.

I'm working in a very old repository that has 75 .csproj files and 9 .sln files. (Which... listen, I'm trying. It's better than it was when I got here.) One of the challenges we've run into is the mixed-mode that .sln files can operate in; some of our .slns are used as workspace configuration for Visual Studio/Rider to limit a developer's view over the repo, and some of them are a build optimization to generate multiple entry-point .dlls as part of a single dependency/build graph. Some of them are both. Because of this, and because our original authors couldn't have predicted the direction our project or the `dotnet` tool would develop, our repository has been swinging between having two to four .sln files at the root, plus a handful nested in other directories.

Omnisharp never really handled this repository well. Admittedly, I don't know what it would look like for a tool to handle this kind of repository _well_, but by selecting a .csproj or .sln file, as johannesergon described, I was able to work quite efficiently in whatever subset of the projects I needed to in order to get my work done.

After my LSP feel over when upgrading to v2.0.320 last week, I was able to restore functionality by rolling back to 1.26.0 and restarting my system. I might try to upgrade again this weekend, or at least find a minimal reproduction for kind of failure I saw, but I'm hesitant to move away from Omnisharp at this time because of how dependent my workflow is on being able to select the .csproj or .sln relevant to my immediate task.

**Comment by jasonmalinowski** [comment]
> For me its a good think to select only the projects which should be loaded, otherwise my my laptop will get super slow. With the omnisharp project/sln selector i can achieve this. I only have to select the correct sln file and wait for omnisharp to load everythink and than i can work normally 🙂

@johannesergon: Just so I can understand a bit more, you have been selecting a solution that does contain all the projects you want, or you don't have that today? Have you been using the selector to regulary switch or do you pick one at the start of your session or go from there?

One thing I realize we don't have documented well but if you do have a solution that you want to use, either set dotnet.defaultSolution as a workspace property which will pick that one on startup (and if it's appropriate for your entire team, check it in), or you can set it to "disable" and we won't load a solution at startup. If you invoke the command pallete and search for "open solution" there is a command to pick the _solution_ that will be loaded. We don't have a picker for individual projects, but if the autodiscovery of a solution didn't pick the right one you can do that.

**Comment by johannesergon** [comment]
@jasonmalinowski 
During the day, I'm switching between our sln files. An sln file contains most of the time only one specific set of projects, and we have like 3-4 sln files which contain more projects (because we are also developing some shared libraries for our specific projects), but we don't have an sln file that contains all of our projects.

FYI:
```
~/workspace$ find . -type f -name "*.sln" | wc -l
26
```

I'm a guy who is developing central components and libraries for our projects; I have the need to search through all of our source code. This is why I (and other guys also) checked out every one of our repositories. VSCode can search a lot better than Azure DevOps 🙂. To test these changes, I'm switching between some projects and including them... Also, if I have some spare time, I'm helping other teams with their work.

I already found this setting `dotnet.defaultSolution` (together with `dotnet-test-explorer.testProjectPath`), but it's not really useful. It's important for me (and my current workflow) to switch between the sln files without switching between different workspace files and/or reloading VSCode.

I hope this is possible with the language server and will not break anything on your side; otherwise, I have to create a new workflow for me. 😅

**Comment by cchhim** [comment]
Please add 

```
{
    "dotnet.defaultSolution": "{YOUR_SOLUTION}.sln"
}
```
In your settings.json file at   .vscode folder. It's working perfectly for me.

**Comment by MrBrax** [comment]
Still broken for me in anything higher than 1.26.0. I'm developing for s&box which uses an sln file in a completely different folder on the drive. My project only has a csproj which isn't even in the root.

**Comment by jasonmalinowski** [comment]
@MrBrax I'm not quite sure what you mean by "it doesn't analyze the game files" -- can you open a separate bug and tag me and we'll investigate from there?

**Comment by jasonmalinowski** [comment]
General update for everyone: last night we uploaded 2.0.356 which adds support for loading all the .csproj in your workspace if there isn't a .sln anywhere to be found. There are of course plenty of workflows this still doesn't cover, but we wanted to get that out quickly to unblock some of the folks here. Right now that version is still in the prerelease channel but we hope to promote it soon.

**Comment by jasonmalinowski** [comment]
> I already found this setting dotnet.defaultSolution (together with dotnet-test-explorer.testProjectPath), but it's not really useful. It's important for me (and my current workflow) to switch between the sln files without switching between different workspace files and/or reloading VSCode.

@johannesergon There is also a "load solution" command you can invoke. If you open the command palate (by default Ctrl+Shift+P on Windows) and search for "open solution" there's a command that will let you pick the other solution file. It's a bit strange right now in picking another one doesn't close the previous solution, so if you have A.sln that has some set of projects and then you also load B.sln, we'll be processing both projects. This may not be perfect for your workflow but I hope that brings it pretty close.

**Comment by MrBrax** [comment]
> @MrBrax I'm not quite sure what you mean by "it doesn't analyze the game files" -- can you open a separate bug and tag me and we'll investigate from there?

In the current docs (that use omnisharp) it is instructed to enable roslyn analyzers or it won't get the symbols required for all the base classes and whatnot. I'm not sure if to open the issue here or for the game, considering it is invite-only and in development.

**Comment by jasonmalinowski** [comment]
@MrBrax File a separate issue here (or create a discussion) so we can discuss this further. There may not be much to discuss, but it'd be good to keep the conversation separate than this one.

**Comment by arkalyanms** [comment]
[2.0.357](https://github.com/dotnet/vscode-csharp/releases/tag/v2.0.357) added release version support for opening projects without a solution file and other tweaks requested in feedback. Please let us know if there are still lingering issues for us to track.

Note: A known feature gap that we are working on is support for legacy framework projects tracked at https://github.com/dotnet/vscode-csharp/issues/5721.

**Comment by johannesergon** [comment]
> > I already found this setting dotnet.defaultSolution (together with dotnet-test-explorer.testProjectPath), but it's not really useful. It's important for me (and my current workflow) to switch between the sln files without switching between different workspace files and/or reloading VSCode.
> 
> @johannesergon There is also a "load solution" command you can invoke. If you open the command palate (by default Ctrl+Shift+P on Windows) and search for "open solution" there's a command that will let you pick the other solution file. It's a bit strange right now in picking another one doesn't close the previous solution, so if you have A.sln that has some set of projects and then you also load B.sln, we'll be processing both projects. This may not be perfect for your workflow but I hope that brings it pretty close.

looks good for me, thank you 🙂👍
but there is one little thing which is not working well for me, everytime i change my solution, my workspace file will be modified (`dotnet.defaultSolution` will be set to a new value), so everybody (who is using this workspace file) have to remember to not commit this change... is there a way to avoid this?

**Comment by jasonmalinowski** [comment]
@johannesergon Is that with DevKit installed?

**Comment by johannesergon** [comment]
> @johannesergon Is that with DevKit installed?

Yes, currently, I'm using `ms-dotnettools.csharp@v2.0.366` (Pre-Release) and `ms-dotnettools.csdevkit@v0.4.2` (Pre-Release) because the C# extension recommends the C# DevKit extension.

And when I'm executing the command `.NET: Open Solution` or `.NET: Close Solution`, this property `dotnet.defaultSolution` will be changed.

**Comment by sharwell** [comment]
> I'll switch back to visual studio even though it's slow as molasses.

Have you filed a performance report for Visual Studio? Our team has a very high close rate for issues filed with these steps:
https://learn.microsoft.com/en-us/visualstudio/ide/how-to-increase-chances-of-performance-issue-being-fixed?view=vs-2019#slowness-and-high-cpu-issues

**Comment by MrBrax** [comment]
> > I'll switch back to visual studio even though it's slow as molasses.
> 
> Have you filed a performance report for Visual Studio? Our team has a very high close rate for issues filed with these steps: https://learn.microsoft.com/en-us/visualstudio/ide/how-to-increase-chances-of-performance-issue-being-fixed?view=vs-2019#slowness-and-high-cpu-issues

No, but it's just a "it's how it is" issue, it's been the case for well over a decade if not two. It's just slow in general due to how heavy of an application it is.

Currently using Rider but it also has some drawbacks, hoping this plugin will get working again in the future as i prefer vscode over all other editors.

**Comment by jasonmalinowski** [comment]
> And when I'm executing the command `.NET: Open Solution` or `.NET: Close Solution`, this property `dotnet.defaultSolution` will be changed.

OK, @johannesergon I've filed https://github.com/microsoft/vscode-dotnettools/issues/399 to track that.

**Comment by jasonmalinowski** [comment]
I'm going to close this bug as I _believe_ we've made enough improvements (or explained things!) to unblock everybody who is in this conversation. If I've missed your comment or your workflow is still blocked, do please let me know, but let's open a new bug to track since this one is getting a bit long.

To summarize, we've got the following changes in or clarified the following bits:

1. If you don't have any solution files, we'll now load the project files directly.
2. If you have more than one solution, instead of doing nothing silently we'll prompt to launch the picker. (This is merged but not shipped yet.)
3. Folks now know about `dotnet sln` and how to create solution files, so if you need to specify a custom set of projects to load, you can use those. At this point, having multiple solution files would be our suggestion for those who need different subsets of projects loaded on a regular basis.
4. Documentation better explaining our selection logic (and pointers to the `dotnet sln` docs) is on its way, since I think one thing we learned here was you all can't read our minds.

Thank you all for your feedback here especially explaining different workflows that we didn't anticipate. If you have a workflow that still isn't covered at all, or have other suggestions, please do file a new bug and feel free to tag me. We're still going to make more improvements here but at this point I think it's best to get new conversations now that this conversation has gone past 50 comments. :smile:

**Comment by govpowerio** [comment]
this is so dumb lmao

**Comment by jasonmalinowski** [comment]
@CarsonKompon The reason I requested new issues is simply because people who are having problem with IntelliSense (for entirely unrelated reasons!) see the title for this issue, go "IntelliSense is gone" and comment here, and then we get to tease all the issues back apart. There are a huge number of systems that all have to work to get Intellisense to work in the end and if those go awry the symptom is often the same. I've also had plenty of cases where a customer thinks an issue is already filed for what they're experiencing, only to discover once we fix it, it wasn't and had they spoken up earlier we could have gotten that fixed sooner.

To confirm for your issue though, "with an existing sln", is the problem again we're picking up the solution when we shouldn't be, or the solution file does contain the projects you expect to load, and you're still not getting intellisense?

**Comment by stevephary** [comment]
its January and the issue is still there. Cant work because of this!!
I have a solution and everything but the problem stills persist, tried uninstall and install but no changes

**Comment by jasonmalinowski** [comment]
@stevephary Can you file a new bug and attach logs as requested in the bug? You can also file a bug from the extension itself which will attach some information automatically. Feel free to tag me so I can take a look.

**Comment by NicoRuedaA** [comment]
Same issue as new user with VSC. Not a good start to learn C#

**Comment by erwinkramer** [comment]
It seems we need to install [IntelliCode for C# Dev Kit](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.vscodeintellicode-csharp) now.

**Comment by jasonmalinowski** [comment]
@erwinkramer That definitely shouldn't be necessary (unless you want it of course!) Can you file a new bug for us to investigate?
